generic.go

```diff
@@ -95,21 +74,75 @@ func NewGeneric(config *Config, queue workqueue.RateLimitingInterface, workers i
 				Help:      fmt.Sprintf("Histogram measuring the time it took to process a %s", &groupKind),
 			},
 		)
+		// Extra api data
+		requestCount := prometheus.NewCounterVec(
+			prometheus.CounterOpts{
+				Name: fmt.Sprintf("processed_%s_count", objectName),
+				Help: fmt.Sprintf("Cumulative number of %s processed", &groupKind),
+			},
+			[]string{"url", "method", "status"},
+		)
+		requestTime := prometheus.NewHistogram(
+			prometheus.HistogramOpts{
+				Name: fmt.Sprintf("processed_%s_time", objectName),
+				Help: fmt.Sprintf("Number of seconds each request to %s takes", &groupKind),
+			},
+		)
+
 		allMetrics := []prometheus.Collector{
 			objectProcessCount, objectProcessTime,
+			requestCount, requestTime,
 		}
 		for _, metric := range allMetrics {
 			if err := constructorConfig.Registry.Register(metric); err != nil {
 				return nil, errors.WithStack(err)
 			}
 		}
 
-		holders[descr.Gvk] = Holder{
-			Cntrlr:             iface,
-			ZapNameField:       descr.ZapNameField,
-			ReadyForWork:       readyForWork,
-			objectProcessCount: objectProcessCount,
-			objectProcessTime:  objectProcessTime,
+		constructed, err := constr.New(
+			constructorConfig,
+			&Context{
+				ReadyForWork: func() {
+					close(readyForWork)
+				},
+				Middleware:  addMiddleware(requestCount, requestTime),
+				Informers:   informers,
+				Controllers: controllers,
+				WorkQueue:   queueGvk,
+			},
+		)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to construct controller for GVK %s", descr.Gvk)
+		}
+
+		inf, ok := informers[descr.Gvk]
+		if ok {
+			inf.AddEventHandler(&GenericHandler{
+				Logger:       gvkLogger,
+				WorkQueue:    queueGvk,
+				ZapNameField: descr.ZapNameField,
+			})
+		}
+
+		if constructed.Interface != nil {
+			controllers[descr.Gvk] = constructed.Interface
+
+			holders[descr.Gvk] = Holder{
+				Cntrlr:             constructed.Interface,
+				ZapNameField:       descr.ZapNameField,
+				ReadyForWork:       readyForWork,
+				objectProcessCount: objectProcessCount,
+				objectProcessTime:  objectProcessTime,
+			}
+		}
+
+		if constructed.Server != nil {
+			servers[descr.Gvk] = ServerHolder{
```

Comment:
There is a check for controllers (should be moved into the `if` above) but no check for `servers` for a GVK. Should be inside of this `if`.